server/services/streaming_sequence_service.go

```diff
@@ -0,0 +1,178 @@
+			name,
+		)
+	}
+	seq := i.(*pb.StreamingSequence)
+
+	i, _ = s.streamingreports.Load(streaming_report(name))
+	rep, _ := i.(*pb.StreamingSequenceReport)
+
+	// Get the number of attempts, which coincides with this attempt's number.
+	n := len(rep.Attempts)
+
+	// Prepare the attempt response defined by the Sequence.
+	st := status.New(codes.OK, "Successful attempt")
+	sendStatusAtIndex := 0
+	var delay time.Duration
+	responses := seq.GetResponses()
+	content := strings.Fields(seq.GetContent())
+	if l := len(responses); l > 0 && n < l {
+		resp := responses[n]
+		delay = resp.GetDelay().AsDuration()
+		st = status.FromProto(resp.GetStatus())
+		sendStatusAtIndex = int(resp.SendStatusAtIndex) - len(strings.Fields(s.sent_content))
+
+		if s.sent_content != "" {
+			words_written := strings.Fields(s.sent_content)
+			content = content[len(words_written):]
+		}
+
+	} else if n > l {
+		st = status.New(codes.OutOfRange, "Attempt exceeded predefined responses")
+	}
+
+	for idx, word := range content {
+		if idx >= sendStatusAtIndex {
+			break
+		}
+		s.sent_content += word + " "
+		err := stream.Send(&pb.AttemptStreamingSequenceResponse{Content: word})
+		if err != nil {
+			return err
+		}
+	}
+
+	echoStreamingHeaders(stream)
+
+	echoStreamingTrailers(stream)
+
+	time.Sleep(delay)
+
+	// Calculate the perceived delay since the last RPC attempt.
+	attDelay := &duration.Duration{}
+	if n > 0 {
+		prev := rep.GetAttempts()[n-1]
+		respTime := prev.GetResponseTime()
+		d := received.Sub(respTime.AsTime())
+		attDelay = ptypes.DurationProto(d)
+	}
+
+	// Clock the time that the server is sending the response
+	responseTime := time.Now()
+	rpb, err := ptypes.TimestampProto(responseTime)
+	if err != nil {
+		return status.Errorf(
+			codes.Internal,
+			err.Error(),
+		)
+	}
+
+	rep.Attempts = append(rep.Attempts, &pb.StreamingSequenceReport_Attempt{
+		AttemptNumber: int32(n),
+		ResponseTime:  rpb,
+		AttemptDelay:  attDelay,
+		Status:        st.Proto(),
+		ContentSent:   s.sent_content,
+	})
+
+	if n+1 >= len(responses) {
+		s.sent_content = ""
+	}
+
+	return st.Err()
+
+}
+
+func (s *sequenceServerImpl) GetStreamingSequenceReport(ctx context.Context, in *pb.GetStreamingSequenceReportRequest) (*pb.StreamingSequenceReport, error) {
+	name := in.GetName()
+	if name == "" {
+		return nil, status.Errorf(
+			codes.InvalidArgument,
+			"The field `name` is required.")
+	}
+
+	report, ok := s.streamingreports.Load(name)
+	if !ok {
+		return nil, status.Errorf(
+			codes.NotFound,
+			"The sequence report with %q does not exist.",
+			name,
+		)
+	}
+
+	return report.(*pb.StreamingSequenceReport), nil
+}
+
+func clone_streamingSequence(s *pb.StreamingSequence) *pb.StreamingSequence {
+	r := make([]*pb.StreamingSequence_Response, len(s.GetResponses()))
+	copy(r, s.GetResponses())
+
+	return &pb.StreamingSequence{
+		Name:      s.GetName(),
+		Content:   s.GetContent(),
+		Responses: r,
+	}
+}
+
+func streaming_report(n string) string {
```

Comment:
Drive by review: `streamingReport` or `streamingReportName`

Do not use underscores in function names.

Reply:
Ack, will correct in both spots